docs/dev/guidelines/server.rst

```diff
@@ -0,0 +1,148 @@
+* RestControllers are by default singletons.
+* RestControllers should not execute business logic but rely on delegation.
+* RestControllers should deal with the HTTP layer of the application.
+* RestControllers should be oriented around a use-case/business-capability.
+
+Route naming conventions:
+
+* Always use kebab-case (e.g. "/exampleAssessment" → "/example-assessment").
+* The routes should follow the general structure entity > entityId > sub-entity ... (e.g. "/exercises/{exerciseId}/participations").
+* Use plural for route's entities.
+* Specify the key entity at the end of the route (e.g. "text-editor/participations/{participationId}" should be changed to "participations/{participationId}/text-editor").
+* Never specify an id that is used only for consistency and not used in the code (e.g. GET "/courses/{courseId}/exercises{exerciseId}/participations/{participationId}/submissions/{submissionId}" can be simplified to GET "/submissions/{submissionId}" because all other entities than the submission are either not needed or can be loaded without the need to specify the id).
+
+Additional notes on the controller methods:
+
+* POST should return the newly created entity
+* Always use @PreAuthorize tag to only allow certain roles to access the method.
+* Never trust user input and check if the passed data exists in the database.
+* Always use different response status codes to notify the client about errors on the server:
+    * Forbidden - the user is not authorized to access the controller.
+    * Bad Request - the request was wrong.
+    * Not Found - can't find the requested data or it should be not accessible yet.
+
+12. Dependency injection:
+=========================
+
+* Some of you may argue this, but by favoring constructor injection you can keep your business logic free from Spring. Not only is the @Autowired annotation optional on constructors, you also get the benefit of being able to easily instantiate your bean without Spring.
+* Use setter based DI only for optional dependencies.
+* Avoid circular dependencies, try constructor and setter based DI for such cases.
+
+13. REST best practices:
+========================
+
+* Verify that API endpoints perform appropriate authorization and authentication consistent with the rest of the code base.
+* Check for other common weaknesses, e.g., weak configuration, malicious user input, missing log events, etc.
+* Handle exceptions and errors with a standard response. Errors are very important in REST APIs. They inform clients that something went wrong, after all.
+
+14. Keep it simple and stupid:
+==============================
+
+* Don’t write complex code.
+* Don’t write code when you are tired or in a bad mood.
+* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because time and resources that will be spent on hard readable code will be much higher than what you get from optimization.
```

Comment:
```suggestion
* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because the time and resources spent on hard-to-read code cost much more than what we gain through optimization.
```